# № 1

Приведём $- x_{1}^{2} - 12 x_{1} x_{2} + 6 x_{1} x_{3} + x_{2}^{2} \left(4 b - 6\right) + x_{2} x_{3} \left(12 b + 12\right) + x_{3}^{2} \left(9 b + 8\right)$ к каноничному виду методом Лагранжа:

Выделим полный квадрат с $x_{1}$:
$-(x_{1} + 6 x_{2} - 3 x_{3})^{2} = - x_{1}^{2} - 12 x_{1} x_{2} + 6 x_{1} x_{3} - 36 x_{2}^{2} + 36 x_{2} x_{3} - 9 x_{3}^{2}$  
В квадратичной форме останется:

$4 b x_{2}^{2} + 12 b x_{2} x_{3} + 9 b x_{3}^{2} + 30 x_{2}^{2} - 24 x_{2} x_{3} + 17 x_{3}^{2} = x_{2}^{2} (4 b + 30) + x_{2} x_{3}(12 b  - 24 ) + x_{3}^{2} \left(9 b + 17\right)$

Если $4b+30 \neq 0$ : выделим полный квадрат: $\left(4 b + 30\right) \left(x_{2} + \frac{x_{3} \left(12 b - 24\right)}{8 b + 60}\right)^{2} = x_{2}^{2} \left(4 b + 30\right) + x_{2} x_{3} \left(12 b - 24 \right) + 9 x_{3}^{2}\frac{b^{2} - 4 b + 4}{4 b^{2} + 60 b + 225}$

Останется: $\left(9 b - \frac{9 \left(b - 2\right)^{2}}{\left(2 b - 15\right)^{2}} + 17\right)x_{3}^{2}$  
Если $b = -\frac{15}{2}$, выделим полный квадрат с $x_{3}$:

$- \frac{101 \left(\frac{57 x_{2}}{101} + x_{3}\right)^{2}}{2} = - \frac{3249 x_{2}^{2}}{202} - 57 x_{2} x_{3} - \frac{101 x_{3}^{2}}{2}$

Таким образом, в каноничной форме при $b \neq -\frac{15}{2}$ квадратичная форма принимает вид:

$y_1^2 + y_2^2 + \left(9 b - \frac{9 \left(b - 2\right)^{2}}{\left(2 b - 15\right)^{2}} + 17\right)y_{3}^{2}$

Иначе, $y_1^2 - \frac{3249 y_{2}^{2}}{202} + y_3^2$

# № 2 

In [1]:
from LinAlgFuncs import *

In [2]:
def argmax(l):
    return max(zip(l, range(len(l))))[1]

def canon_square_Lagrange(Q):
    assert len(Q.shape()) == 2, "Need 2D Matrix"
    n = Q.shape()[0]
    Q = Q.copy()
    res = I(n)
    assert n == Q.shape()[1], "Need aquare matrix"
    i = 0
    while i<n:
        S = I(n)
        if Q[i][i] != 0 and max(map(abs, Q[i][i:])) != 0:
            S[i][i+1:] = -1*Q[i][i+1:]*Fraction(1,Q[i][i])
            i+=1
        elif max(map(abs, Q[i][i:])) != 0:
            t = argmax(map(abs, Q[i][i:]))
            S[i][t] = -1
            S[t][i] = 1
        else:
            i+=1
        Q = dot(S.transpose2(),dot(Q, S))
        res = dot(res,S)
    return res

In [3]:
G = tensor_from_iterable([
    [33, 2 , 17],
    [ 2, 22, -9],
    [17, -9, 22]
])

In [4]:
C = canon_square_Lagrange(G)
print("матрица перехода: \n", C)
G1 = dot(C.transpose2(), dot(G, C))
G1

матрица перехода: 
 [[       1,    -2/33, -196/361],
 [       0,        1,  331/722],
 [       0,        0,        1]]


[[      33,        0,        0],
 [       0,   722/33,        0],
 [       0,        0, 6241/722]]

Все элементы на диагонали - неотрицательны $\Rightarrow G_1$ - матрица Грамма $\Rightarrow G$ - матрица Грамма    
На диагонали матрицы $G_1$ записаны квадраты длин векторов ортогональной системы, в качестве ортогональной системы векторов возьмём стандартный базис в $\mathbb{R}^3$:  
$[1,0,0]$,   
$[0,1,0]$,  
$[0,0,1]$  
Умножим их на длины векторов:  
$v_1 =\left[\sqrt{33}, 0, 0\right]$  
$v_2 = \left[0, \frac{19 \sqrt{66}}{33}, 0\right]$  
$v_3 = \left[0, 0, \frac{79 \sqrt{2}}{38}\right]$

Дальше умножим матрицу $A = (v_1, v_2, v_3) = \begin{pmatrix}
\sqrt{33} & 0 & 0\\
0 & \frac{19 \sqrt{66}}{33} & 0\\
0 & 0 & \frac{79 \sqrt{2}}{38}
\end{pmatrix}$ справа на обратную матрицу перехода:  

In [5]:
inverse_Gauss(C)

[[       1,     2/33,    17/33],
 [       0,        1, -331/722],
 [       0,        0,        1]]

$A \times C^{-1} = \begin{pmatrix}
\sqrt{33} & \frac{2 \sqrt{33}}{33} & \frac{17 \sqrt{33}}{33}\\
0 & \frac{19 \sqrt{66}}{33} & - \frac{331 \sqrt{66}}{1254}\\
0 & 0 & \frac{79 \sqrt{2}}{38}
\end{pmatrix}$

Соотвтетственно вектора =  
$\left[\sqrt{33} , 0 , 0\right]$  
$\left[\frac{2 \sqrt{33}}{33} , \frac{19 \sqrt{66}}{33} , 0\right]$  
$\left[\frac{17 \sqrt{33}}{33} , - \frac{331 \sqrt{66}}{1254} , \frac{79 \sqrt{2}}{38}\right]$

# № 3 

Чтобы дополнить систему векторов до ортонормированной, добавим к ней единичный базис и ортогонализуем:

In [6]:
v = Tensor([-2,-8,-7,-2])*Fraction(1,11)
e = I(4)
e.insert(0, v)
ort = orthogonalize(e)
ort

[[-2/11, -8/11, -7/11, -2/11],
 [117/121, -16/121, -14/121,  -4/121],
 [      0,  53/117, -56/117, -16/117],
 [     0,      0,   4/53, -14/53]]

In [7]:
print(Tensor(ort).__repr__(repr_items=1))

[[  Fraction(-2, 11),   Fraction(-8, 11),   Fraction(-7, 11),   Fraction(-2, 11)],
 [Fraction(117, 121), Fraction(-16, 121), Fraction(-14, 121),  Fraction(-4, 121)],
 [    Fraction(0, 1),  Fraction(53, 117), Fraction(-56, 117), Fraction(-16, 117)],
 [    Fraction(0, 1),     Fraction(0, 1),    Fraction(4, 53),  Fraction(-14, 53)]]


Затем, отнормируем

In [8]:
print(dot(ort[0], ort[0])) # - этот вектор отнормирован

1


In [9]:
print(dot(ort[1], ort[1])) # - этот вектор не отнормирован

117/121


$\left[\frac{117}{121}, -\frac{16}{121}, -\frac{14}{121}, -\frac{4}{121}\right]\frac{11}{\sqrt{117}} = \ 
\left[\frac{3 \sqrt{13}}{11}, - \frac{16 \sqrt{13}}{429}, - \frac{14 \sqrt{13}}{429}, - \frac{4 \sqrt{13}}{429}\right]$

In [10]:
print(dot(ort[2],ort[2])) # - этот вектор не отнормирован

53/117


$\left[0, \frac{53}{117}, - \frac{56}{117}, - \frac{16}{117}\right]*\sqrt{\frac{117}{53}} = \left[0, \frac{\sqrt{689}}{39}, - \frac{56 \sqrt{689}}{2067}, - \frac{16 \sqrt{689}}{2067}\right]$

In [11]:
print(dot(ort[3],ort[3])) # - этот вектор не отнормирован

4/53


$\left[0, 0, \frac{4}{53}, - \frac{14}{53}, \right] * \sqrt{\frac{53}{4}} = \left[0, 0, \frac{12 \sqrt{689}}{2809}, - \frac{42 \sqrt{689}}{2809}, \right]$

# № 4

In [12]:
U_sleq = tensor_from_iterable([
    [-7, -3, 7, 3],
    [2, 1, -2, -1]
])

In [13]:
U_basis = solve_hsle(U_sleq)
U_basis

[[1, 0, 1, 0],
 [0, 1, 0, 1]]

Найдём проекцию по формуле:
$pr_Uv = A(A^TA)^{-1}A^Tv$ 

In [14]:
v = Tensor([3,-3,5,-1])

In [15]:
A = U_basis
pr_U_v = dot(dot(dot(A.transpose2(),inverse_Gauss(dot(A, A.transpose2()))),A),v)
pr_U_v

[[ 4],
 [-2],
 [ 4],
 [-2]]

$pr_Uv + ort_Uv = v \iff ort_Uv = v-pr_Uv$

In [16]:
ort_v = v - pr_U_v.transpose2()[0]
ort_v

[-1, -1,  1,  1]

Расстояние от $v$ до $U$ = расстояние от вектора $v$ до его проекции = длина его ортогональной состовляющей :$\sqrt{(ort_Uv, ort_Uv)}; (ort_Uv, ort_Uv) = $

In [17]:
print(dot(ort_v, ort_v))

4


Итого расстояние = $\sqrt{4} = 2$